pkg/executor/join/semi_join_probe.go

```diff
@@ -0,0 +1,332 @@
+	err = s.concatenateProbeAndBuildRows(joinedChk, sqlKiller, true)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if s.unFinishedProbeRowIdxQueue.IsEmpty() && s.offset == s.chunkRows {
+			// To avoid `Previous chunk is not probed yet` error
+			s.currentProbeRow = s.chunkRows
+		}
+	}()
+
+	if joinedChk.NumRows() > 0 {
+		s.selected = s.selected[:0]
+		s.selected, err = expression.VectorizedFilter(s.ctx.SessCtx.GetExprCtx().GetEvalCtx(), s.ctx.SessCtx.GetSessionVars().EnableVectorizedExpression, s.ctx.OtherCondition, chunk.NewIterator4Chunk(joinedChk), s.selected)
+		if err != nil {
+			return err
+		}
+
+		s.setIsMatchedRows()
+	}
+
+	if s.unFinishedProbeRowIdxQueue.IsEmpty() {
+		for remainCap > 0 && (s.offset < s.chunkRows) {
+			rowNumToAppend := int(math.Min(float64(remainCap), float64(s.chunkRows-s.offset)))
+
+			for index, usedColIdx := range s.lUsed {
+				dstCol := chk.Column(index)
+				srcCol := s.currentChunk.Column(usedColIdx)
+				chunk.CopySelectedRowsWithRowIDFunc(dstCol, srcCol, s.isMatchedRows, s.offset, s.offset+rowNumToAppend, func(i int) int {
+					return i
+				})
+			}
+
+			s.offset += rowNumToAppend
+			remainCap -= rowNumToAppend
+		}
+	}
+	return
+}
+
+func (s *semiJoinProbe) probeForRightSideBuildNoOtherCondition(chk *chunk.Chunk, remainCap int, sqlKiller *sqlkiller.SQLKiller) (err error) {
+	meta := s.ctx.hashTableMeta
+	tagHelper := s.ctx.hashTableContext.tagHelper
+
+	if cap(s.offsets) == 0 {
+		s.offsets = make([]int, 0, remainCap)
+	}
+
+	s.offsets = s.offsets[:0]
+	matched := false
+
+	for remainCap > 0 && s.currentProbeRow < s.chunkRows {
+		if s.matchedRowsHeaders[s.currentProbeRow] != 0 {
+			candidateRow := tagHelper.toUnsafePointer(s.matchedRowsHeaders[s.currentProbeRow])
+			if isKeyMatched(meta.keyMode, s.serializedKeys[s.currentProbeRow], candidateRow, meta) {
+				s.matchedRowsHeaders[s.currentProbeRow] = 0
+				matched = true
+				remainCap--
+			} else {
+				s.probeCollision++
+				s.matchedRowsHeaders[s.currentProbeRow] = getNextRowAddress(candidateRow, tagHelper, s.matchedRowsHashValue[s.currentProbeRow])
+			}
+		} else {
+			if matched {
+				s.offsets = append(s.offsets, s.usedRows[s.currentProbeRow])
+				matched = false
+			}
+			s.currentProbeRow++
+		}
+	}
+
+	err = checkSQLKiller(sqlKiller, "killedDuringProbe")
+	if err != nil {
+		return err
+	}
+
+	if s.currentProbeRow < s.chunkRows && matched {
+		s.offsets = append(s.offsets, s.usedRows[s.currentProbeRow])
+	}
+	generateResultChk(chk, s.currentChunk, s.lUsed, s.offsets)
+	return
+}
+
+func generateResultChk(resultChk *chunk.Chunk, probeChk *chunk.Chunk, lUsed []int, offsets []int) {
+	for index, colIndex := range lUsed {
+		srcCol := probeChk.Column(colIndex)
+		dstCol := resultChk.Column(index)
+		for _, offset := range offsets {
+			dstCol.AppendCellNTimes(srcCol, offset, 1)
```

Comment:
I think you can add a new function to append all the `offsets` element from srcCol to dstCol

Reply:
And I think you need to update the virtualNumRows